Insert post title instead of URL, when adding a link to an existing post

pwkip opened the pull request:
Fixes #11930
Make the link anchor also retrieve the title of the post/page

## Description
- New behavior: If you insert a new link to an existing post via the inline-link modal, the post title is automatically inserted as link text.
- Previous behavior: If you insert a new link to an existing post via the inline-link modal, the actual URL is inserted as the link text.

## How has this been tested?
Tested on local development server. Tested the following cases:
- [x] insert new link with modal --> start typing --> select post from list. Result: Post title is inserted as link text
- [x] select some text --> insert link with modal --> link remains unchanged
- [x] insert new link with modal --> insert a link to an extarnal page --> The actual URL is inserted as link text

## Screenshots
![Fix for issue 11930](https://bdwm.be/wp-content/uploads/2020/03/11930.gif)

## Types of changes
New feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. (I think the code is obvious by itself) <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

pwkip committed:
Fix for #11930
Make the link anchor also retrieve the title of the post/page

aduth commented:
This seems like a sensible change to me. I know there's been some history with the suggestion title and wanting to avoid some of its applications (#19735, #19739), but in terms of a default text for a new link, it seems like a good option to use as a default. My only concern would be if a manually entered URL (as in a user pasting a URL to the input field) would be selected and have an associated title, it's not clear to me if the user would expect the title to be used, or the URL. I don't know that it's currently the case that a title is assigned in these instances, however.

aduth commented on packages/format-library/src/link/inline.js:
```diff
@@ -119,8 +119,13 @@ function InlineLinkUI( {
 
 		if ( isCollapsed( value ) && ! isActive ) {
+			const newText =
+				nextValue.hasOwnProperty( 'title' ) && nextValue.title !== ''
```
Do you see any benefit of this implementation vs...

```suggestion
				nextValue.title
```

?

Or even simplified further:

```js
const newText = nextValue.title || newUrl;
```

pwkip replied:
That seems to work just as well. Looks a lot cleaner! I noticed the title property will not always exist inside nextValue and my PHP mind was thinking this would throw an error, but seems like in JS nextValue.jibberish will just return undefined, which is falsy. So yeah, this will work just the same. Will update the PR.

aduth replied:
Yeah, the original implementation could make sense if you needed to differentiate `undefined` or `''` (empty string), but as you mention, it should be fine enough to simply use falsyness as an indicator.

aduth commented:
There's unfortunately not any useful unit tests associated with this code. However, there is a pretty exhaustive suite of end-to-end tests for links:

https://github.com/WordPress/gutenberg/blob/master/packages/e2e-tests/specs/editor/various/links.test.js

It would be good if we could account for this new behavior somewhere in those tests.

Depending on whether you're using the default development environment, it could be a little bit of extra work to get the end-to-end tests running in your environment:

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/testing-overview.md#end-to-end-testing

Is that something you might be interested in taking on? Let me know if I can help.

getdave reviewed (commented):
Thank for submitting this PR. It's very much appreciated 👍 

I agree with @aduth that this seems like a sensible change. We should be aware of edge cases and try and include these in the e2e tests he references.

I did spot one bug. If you create some Pages with long titles then when the link is created only part of the `title` gets marked as the hyperlink. I'm not sure if that's a bug with this or a wider bug that exists in `master`.

![Screen Capture on 2020-03-31 at 16-53-00](https://user-images.githubusercontent.com/444434/78047534-610d1480-7370-11ea-91ac-8777fb16ec7f.gif)

pwkip committed:
Fix long titles
Also made code a bit more readable

pwkip commented:
> I did spot one bug. If you create some Pages with long titles then when the link is created only part of the `title` gets marked as the hyperlink. I'm not sure if that's a bug with this or a wider bug that exists in `master`.

@getdave Nice find. fixed it. Looks like the formatting was still applied to the length of the URL, I changed it to the length of the inner text now. Also did a quick find for other places where the newUrl was used. I think we're good now.

@aduth I'm using the default `npm run dev` command for the development server. I'll look into the links you shared and see if I can figure out what I need to do.

pwkip commented:
@aduth I might need some help. All tests keep failing with this message:
["Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in %s.%s the componentWillUnmount method
![e2e-error](https://user-images.githubusercontent.com/1150196/78071958-99c3e280-739e-11ea-82b3-0f7b999ba04f.png)

I'll look into the error tomorrow. Probably something stupid. Maybe you know a quick fix?

aduth commented:
@pwkip I believe this issue may be because the built copy of Gutenberg is generated from `npm run dev`, and because there are presently many warnings in development mode of the editor, the tests fail, because there are mechanisms in place which ensure that no `console.error` or `console.warn` is allowed to occur in the end-to-end test environment:

Related:

- #17355
- #14845
- https://github.com/WordPress/gutenberg/blob/981a45db76a0351cbf9835ad03f801e6e6fe466b/packages/e2e-tests/config/setup-test-framework.js#L117-L206

Short-term, you can probably resolve this by running `npm run build` (the production build) before you run the end-to-end tests, or simply ignore those specific failures, since they'd not be considered by this Travis build anyways.

pwkip committed:
Add test to check link to existing post

pwkip commented:
@aduth I've created a new test `'will use Post title as link text if link to existing post is created without any text selected'` and also added it to the pull request. Not 100% sure I did it right, would appreciate a quick review of that as well.

pwkip committed:
small change in naming

aduth commented on packages/e2e-tests/specs/editor/various/links.test.js:
```diff
@@ -26,6 +26,42 @@ describe( 'Links', () => {
+
+		const actualText = await page.evaluate(
+			() =>
+				document.querySelector( '.block-editor-rich-text__editable a' )
+					.textContent
```
It's only a small point and probably won't make much of a difference at least in the short-term, but this is a pretty broad selector. While it might not seem like it, there's often many instances of block elements and/or rich text elements in a given page, so the fact that we hope the first one we match is the one we expect it to be could potentially bite us in the future if those circumstances were to change.

Generally, it is quite hard to pinpoint the specific block we're looking for. 

A few approaches I've seen and used in the past:

- A selector which identifies blocks within the block list, then picks the one at the expected index
- Selecting from where we expect focus to be, if focus is assumed to return to the field after the link is applied
   - Something like `await page.evaluate( () => document.activeElement.closest( '.block-editor-rich-text__editable a' ) );`

pwkip replied:
I had a pretty hard time figuring out the test, so I was happy when I finally saw mine succeed 😄. I'll keep your suggestions in mind for any future tests

aduth reviewed (approved):
This looks and works great! Thanks, and congratulations on your first contribution! 🎉

pwkip commented:
@aduth Thanks for the follow up and all your helpful feedback! I learned a lot and hope to contribute more in the future. 👍

getdave commented:
Congratulations on your PR acceptance @pwkip!
